### PR TITLE
fix: simplify reading

### PR DESCRIPTION
**Pull request by ronag**
I believe this is a simpler, faster and more correct way of doing this, maybe?

Fixes: https://github.com/mcollina/undici/issues/50

### CODE REVIEW

**Comment by ronag** [comment]
> My empiric testing showed that .read() is generally faster than pause()/resume().

I'm sceptical whether that difference is significant enough to warrant the complexity here. Not sure how to benchmark it though.

**Comment by ronag** [comment]
I would like request we do this, at least to reduce the cognitive load until we've ironed out the rest of the issues/bugs.

**Comment by ronag** [comment]
The next problem to solve here will be that the client should keep reading data from the connection if the body is `destroyed`. Which in this implementation would be:

```js
// only pause if we have a valid body
if (body && !body.destroyed && body.push(...)) {
  socket.pause()
}
```

Not sure how to resolve with the existing one.

**Comment by mcollina** [comment]
this now conflicts.

**Comment by ronag** [comment]
I encountered some timing bugs while playing around with this, will try to sort those out together with this.

**Comment by ronag** [comment]
@mcollina: PTAL

File: test/client-errors.js [hunk 1]
```diff
@@ -268,7 +268,7 @@ test('POST which fails should error response', (t) => {
   t.plan(4)
 
   const server = createServer()
-  server.once('request', (req, res) => {
+  server.on('request', (req, res) => {
```
Comment (ronag) [inline test/client-errors.js:271]: This test shouldn't have passed before... some other bug that got resolved with this PR.

File: lib/client.js [hunk 2]
```diff
@@ -394,7 +370,13 @@ function reset (client, err) {
     const skipBody = request.method === 'HEAD'
 
     if (!skipBody) {
-      client[kLastBody] = new client[kStream].Readable({ read: client[kRead] })
+      client[kLastBody] = new client[kStream].Readable({
+        read () {
+          if (client.socket) {
+            client.socket.resume()
+          }
+        }
```
Comment (ronag) [inline lib/client.js:378]: Assuming turbofan can extract and optimize this.

**Comment by ronag** [comment]
flaky test again

File: lib/client.js [hunk 3]
```diff
@@ -204,9 +202,9 @@ class Client extends EventEmitter {
             // TODO we might want to wait before previous in-flight
             // requests are finished before destroying
             if (this.socket) {
-              destroySocket(this, err, cb)
+              destroySocket(this, err, callback)
             } else {
-              cb(err, null)
+              callback(err, null)
```
Comment (ronag) [inline lib/client.js:207]: This was using a non-wrapped callback before

**Comment by mcollina** [comment]
this conflicts now

**Comment by ronag** [comment]
fixed

**Comment by mcollina** [comment]
before lgtm this, can you please check if the benchmarks degrades? https://github.com/mcollina/undici/tree/master/benchmarks

**Comment by ronag** [comment]
master:
benchmarks$ node undici.js 
requests: 21.902s

#51:
benchmarks$ node undici.js 
requests: 22.202s

**Pull request merged**